### DISCUSSION TITLE
How to install and run datasette on freebsd ?

### DISCUSSION THREAD

**Discussion by ghost** [simonw/datasette#2015] [Q&A] [answered]
So far, I have tried datasette on Linux (debian and raspbian) and on Mac OSX with homebrew. In all these cases, I installed datasette with pipx and it just worked. I believe in all these situations python comes with the sqlite module included. But on freebsd (using the freebsd "package" for python3.9), it does not:

```
 2+0 ~$ python3.9
Python 3.9.16 (main, Dec 17 2022, 01:13:33)
[Clang 13.0.0 (git@github.com:llvm/llvm-project.git llvmorg-13.0.0-0-gd7b669b3a on freebsd13
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'sqlite'
>>> import sqlite3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/sqlite3/__init__.py", line 57, in <module>
    from sqlite3.dbapi2 import *
  File "/usr/local/lib/python3.9/sqlite3/dbapi2.py", line 27, in <module>
    from _sqlite3 import *
ModuleNotFoundError: No module named '_sqlite3'
>>>
```

... and this is pretty much what I also get when I run the datasette binary I have installed with pipx (which succeeded without errors or even warnings).

So, how to use datasette in such situations? I really don't want to build my own python, or manually wrangle pip options.

-- 
Ian

**Comment by cldellow** [comment, marked as answer]
You could try installing `pysqlite3-binary` from pip. I don't know if it has pre-built FreeBSD versions, though.

pysqlite3-binary is a package that bundles a pre-compiled, modern sqlite to work around the fact that many distros have a very old SQLite. Datasette is smart enough to use it in preference to the standard sqlite3 module if it's available.

**Reply by ghost** [reply]
I'll try that, thanks.

**Reply by ghost** [reply]
So, for whatever reason, this does not quite  work: `pip install pysqlite3-binary` errors out, reporting no available versions. I ended up doing `pip install pysqlite3` which failed initially because it couldn't find `sqlite3.h` which is in `/usr/local/include`. Finally this worked:

```
CPPFLAGS=-I/usr/local/include pip install pysqlite3
```

Which, of course, is pretty much what I meant by "manually wrangling pip options". :-( But at least it worked, and it wasn't really as bad as I feared. Thanks again for trying to help.

Why can't pip use pkgconfig to set up compilations? That would work on the vast majority of systems, including this one.

-- 
Ian
